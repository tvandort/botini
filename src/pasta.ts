import { parse } from "node-html-parser";

/**
 *
 * @param revision Defaults to revision on 23 November 2024
 */
export async function fetchUncookedPastas(revision = 1259032650) {
  const response = await fetch(
    `https://en.wikipedia.org/api/rest_v1/page/html/List_of_pasta/${revision}?redirect=true`,
  );
  const text = await response.text();
  const root = parse(text);

  const captions = root.querySelectorAll("caption");
  const tables = [];
  for (const caption of captions) {
    tables.push(caption.closest("table"));
  }

  const pastas: { text: string; url: string | undefined }[] = [];
  for (const table of tables) {
    const rows = table?.querySelector("tbody")?.querySelectorAll("tr") ?? [];
    for (let rowIndex = 1; rowIndex < rows.length; rowIndex++) {
      const row = rows[rowIndex];
      pastas.push({
        text: row.querySelectorAll("th")[0].innerText,
        url: row.querySelectorAll("a")[0].getAttribute("href"),
      });
    }
  }

  return pastas;
}

const referenceTag = /\[\d*\]/;
const parenthetical = /\(.*\)/;
const leadingOr = /^or\s/;
const betweenOr = /\sor\s/;

// surely this should be transformUncookedPasta
export function transformRawPasta(pastasTitle: string) {
  const pastas = [];

  if (parenthetical.test(pastasTitle)) {
    pastas.push(pastasTitle.replace(parenthetical, ""));
    const matches = parenthetical.exec(pastasTitle) ?? [];
    for (const match of matches) {
      pastas.push(match);
    }
  } else if (betweenOr.test(pastasTitle)) {
    for (const pasta of pastasTitle.split(betweenOr)) {
      pastas.push(pasta);
    }
  } else {
    pastas.push(pastasTitle);
  }

  return pastas
    .map((pasta) => pasta.replace(referenceTag, ""))
    .map((pasta) => pasta.trim())
    .map((pasta) =>
      pasta.replace("(", "").replace(")", "").replace(leadingOr, ""),
    );
}

export function mapRawPastas(
  pastas: Awaited<ReturnType<typeof fetchUncookedPastas>>,
) {
  const transformed = pastas.map((pasta) => ({
    ...pasta,
    names: transformRawPasta(pasta.text),
  }));

  const allNames = [];
  const dictionary: Record<
    string,
    { names: string[]; text: string; url: string | undefined }
  > = {};
  for (const pasta of transformed) {
    for (const name of pasta.names) {
      allNames.push(name);
      dictionary[name] = pasta;
    }
  }

  return {
    allNames,
    dictionary,
  };
}
