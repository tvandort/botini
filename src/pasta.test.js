import { fetchPastas } from "./pasta";

describe("pastas", () => {
  test("that the pasta list works", async () => {
    const pastas = await fetchPastas();
    expect(pastas).toMatchSnapshot();
  });
});
