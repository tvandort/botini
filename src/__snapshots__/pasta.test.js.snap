// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pastas that the pasta list works 1`] = `
[
  "Barbine",
  "Bavette",
  "Bigoli",
  "Bucatini",
  "Busiate (or busiati)",
  "Capellini",
  "Fedelini",
  "Ferrazzuoli",
  "Fettuccine",
  "Fileja",
  "Lagane[25]",
  "Lasagne",
  "Lasagnette",
  "Lasagnotte",
  "Linguine",
  "Maccheroni alla molinara",
  "Maccheroncini di Campofilone",
  "Mafalde",
  "Matriciani",
  "Pappardelle",
  "Perciatelli",
  "Picagge",
  "Pici",
  "Pillus",
  "Rustiche",
  "Sagne 'ncannulate",
  "Scialatelli or scialatielli",
  "Spaghetti",
  "Spaghetti alla chitarra",
  "Spaghettini",
  "Spaghettoni",
  "Stringozzi",
  "Su Filindeu",
  "Tagliatelle",
  "Tagliolini",
  "Trenette",
  "Tripoline",
  "Vermicelli",
  "Ziti",
  "Anelli",
  "Boccoli",
  "Calamarata",
  "Campanelle or torchio",
  "Cappelli da chef",
  "Casarecce",
  "Cascatelli",
  "Castellane",
  "Cavatappi",
  "Cavatelli",
  "Chifferi",
  "Cicioneddos",
  "Conchiglie",
  "Creste di gallo",
  "Fagioloni",
  "Farfalle",
  "Fazzoletti",
  "Festoni",
  "Fiorentine",
  "Fiori",
  "Fusilli",
  "Fusilli bucati",
  "Garganelli",
  "Gemelli",
  "Gnocchi",
  "Gomiti",
  "Lanterne",
  "Lorighittas",
  "Macaroni",
  "Maccheroncelli",
  "Mafaldine",
  "Maltagliati",
  "Malloreddus",
  "Mandala",
  "Marille",
  "Mezzani",
  "Mezze maniche",
  "Mezze penne",
  "Mezzi bombardoni",
  "Nuvole",
  "Paccheri",
  "Passatelli",
  "Pasta al ceppo",
  "Penne",
  "Penne ricce",
  "Picchiarelli",
  "Pipe rigate",
  "Pizzoccheri",
  "Quadrefiore",
  "Radiatori",
  "Riccioli",
  "Ricciolini",
  "Ricciutelle",
  "Rigatoncini",
  "Rigatoni",
  "Rombi",
  "Rotelle",
  "Sagnette",
  "Sagnarelli",
  "Sedani",
  "Spirali",
  "Spiralini (Scharfalini)",
  "Strapponi",
  "Strozzapreti",
  "Testaroli",
  "Tortiglioni",
  "Treccioni",
  "Trenne",
  "Trofie",
  "Trottole",
  "Tuffoli",
  "Vesuvio",
  "Cencioni",
  "Corzetti",
  "Fainelle",
  "Foglie d'ulivo",
  "Orecchiette",
  "Acini di pepe",
  "Alphabet pasta",
  "Anchellini",
  "Anelli",
  "Anellini",
  "Armonie",
  "Conchigliette",
  "Coquillettes",
  "Coralli",
  "Corallini",
  "Cuscussu",
  "Ditali",
  "Egg barley",
  "Farfalline",
  "Fideos[126]",
  "Filini",
  "Fregula",
  "Funghini",
  "Gianduietta",
  "Grano",
  "Gramigna",
  "Grattini",
  "Grattoni",
  "Margheritine",
  "Merletti",
  "Midolline",
  "Occhi di passero",
  "Occhi di pernice",
  "Orzo",
  "Pastina",
  "Piombi",
  "Ptitim",
  "Puntine",
  "Quadrettini",
  "Sorprese",
  "Stelle",
  "Stortini",
  "Tripolini",
  "Agnolini",
  "Agnolotti",
  "Caccavelle",
  "Cannelloni",
  "Cappelletti",
  "Caramelle",
  "Casoncelli",
  "Casunziei",
  "Conchiglioni",
  "Culurgioni",
  "Fagottini",
  "Lumache",
  "Mezzelune",
  "Occhi di lupo",
  "Pansotti",
  "Pavese agnolotti",
  "Ravioli",
  "Rotolo ripieno",
  "Sacchettoni",
  "Tortelli",
  "Tortellini",
  "Tortelloni",
  "Tufoli",
  "Canederli",
  "Donderet",
  "Gnocchi",
]
`;
